### PR TITLE
Extension of State Machines to support simple nesting

### PR DESCRIPTION
**Pull request by SaMuellerDLR**
* State now extends StateMachine
* Redirecting diagram features for State in Diagram to diagram featurs of StateMachine
* Adapated property checker to check an object inherits from StateMachine instead of checking if it is the category Statemachine to enable opening States in the State machine editor.
* Fixed various minor issues regarding updating and labeling transitions

### CODE REVIEW

**Comment by franzTobiasDLR** [comment]
When I tried it out, I got an error when renaming states (triggering the rename action ether by F2 or by simple mouse click)... Since I never saw that behavior before, could that be related to your changes?
grafik

**Comment by franzTobiasDLR** [comment]
> When I tried it out, I got an error when renaming states (triggering the rename action ether by F2 or by simple mouse click)... The state shape disapeared. When renaming the state from the VirSat Editor and updateing the diagram evything worked fine.
> 
> Since I never saw that behavior before, could it be related to your changes? grafik

Just checked, in development this doesn't happen. Looks like there is a problem in the rename action...

**Comment by SaMuellerDLR** [comment]
Adressed all comments by properly delegating feature calls from State to StateMachine diagram features.

File: de.dlr.sc.virsat.model.extension.statemachines.ui/src/de/dlr/sc/virsat/model/extension/statemachines/ui/diagram/features/StateMachineDiagramFeatureProvider.java [hunk 1]
```diff
@@ -110,18 +113,27 @@ public IAddFeature getAddFeature(IAddContext context) {
 	public ICustomFeature[] getCustomFeatures(ICustomContext context) {
 		PictogramElement[] pe = context.getPictogramElements();
 		Object object = getBusinessObjectForPictogramElement(pe[0]);
-		if (object instanceof State) {
+		List<ICustomFeature> customFeaturs = new ArrayList<>();
```
Comment (pchrszon-dlr) [inline de.dlr.sc.virsat.model.extension.statemachines.ui/src/de/dlr/sc/virsat/model/extension/statemachines/ui/diagram/features/StateMachineDiagramFeatureProvider.java:116]: Rename `customFeaturs` to `customFeatures`?

File: de.dlr.sc.virsat.model.extension.statemachines.ui/src/de/dlr/sc/virsat/model/extension/statemachines/ui/diagram/features/states/StateAddFeature.java [hunk 2]
```diff
@@ -89,64 +93,71 @@ public boolean canAdd(IAddContext context) {
...
-    	// set container shape for direct editing after object creation
-    	directEditingInfo.setMainPictogramElement(containerShape);
-    	// set shape and graphics algorithm where the editor for
-    	// direct editing shall be opened after object creation
-    	directEditingInfo.setPictogramElement(containerShape);
-    	directEditingInfo.setGraphicsAlgorithm(nameText);
-    	
-    	layoutPictogramElement(containerShape);
+		Object target = getBusinessObjectForPictogramElement(targetContainer);	
+		if (target instanceof StateMachine) {
```
Comment (pchrszon-dlr) [inline de.dlr.sc.virsat.model.extension.statemachines.ui/src/de/dlr/sc/virsat/model/extension/statemachines/ui/diagram/features/states/StateAddFeature.java:104]: Inconsistent and confusing indentation, but this was also present before.

**Review by pchrszon-dlr — APPROVED** [review]
I tried out the diagram editor as well as the nesting of state machines and found no issues, everything is working as expected. The previously described problems (disappearing node shapes) have been fixed.

**Comment by franzTobiasDLR** [comment]
Awsome thanks! I found one more minor thing: renaming transitions in nested diagrams doesn't seem to work... Well, I think the renaming process actually works, but the diagram's label is not updated....

File: de.dlr.sc.virsat.model.extension.statemachines.ui/src/de/dlr/sc/virsat/model/extension/statemachines/ui/diagram/features/StateMachineDiagramFeatureProvider.java (hunk 1 above)
Comment (SaMuellerDLR) [reply de.dlr.sc.virsat.model.extension.statemachines.ui/src/de/dlr/sc/virsat/model/extension/statemachines/ui/diagram/features/StateMachineDiagramFeatureProvider.java:116]: Nice catch!

**Comment by SaMuellerDLR** [comment]
> Awsome thanks! I found one more minor thing: renaming transitions in nested diagrams doesn't seem to work... Well, I think the renaming process actually works, but the diagram's label is not updated....

I'm not sure if I reproduced your issue, but I did manage to find an issue when creating transitions in substates (they were added to the parent state machine). Please recheck if I fixed a different issue or the one you observed as well.

**Comment by franzTobiasDLR** [comment]
> > Awsome thanks! I found one more minor thing: renaming transitions in nested diagrams doesn't seem to work... Well, I think the renaming process actually works, but the diagram's label is not updated....
> 
> I'm not sure if I reproduced your issue, but I did manage to find an issue when creating transitions in substates (they were added to the parent state machine). Please recheck if I fixed a different issue or the one you observed as well.

Yes, your change fixed it 👍

**Review by franzTobiasDLR — APPROVED** [review]

**Pull request merged**